examples/android/clientcache/app/src/main/java/io/grpc/clientcacheexample/ClientCacheExampleActivity.java

```
@@ -0,0 +1,150 @@
+    private ManagedChannel mChannel;
+
+    @Override
+    protected void onPreExecute() {
+      mHost = mHostEdit.getText().toString();
+      mMessage = mMessageEdit.getText().toString();
+      String portStr = mPortEdit.getText().toString();
+      mPort = TextUtils.isEmpty(portStr) ? 0 : Integer.valueOf(portStr);
+      mResultText.setText("");
+    }
+
+    @Override
+    protected String doInBackground(Void... nothing) {
+      try {
+        mChannel = ManagedChannelBuilder.forAddress(mHost, mPort).usePlaintext(true).build();
+        Channel channelToUse =
+            ClientInterceptors.intercept(
+                ManagedChannelBuilder.forAddress(mHost, mPort).usePlaintext(true).build(),
+                SafeMethodCachingInterceptor.newSafeMethodCachingInterceptor(cache));
+        HelloRequest message = HelloRequest.newBuilder().setName(mMessage).build();
+        HelloReply reply;
+        if (getCheckBox.isChecked()) {
+          MethodDescriptor<HelloRequest, HelloReply> safeCacheableUnaryCallMethod =
+              GreeterGrpc.METHOD_SAY_HELLO.toBuilder().setSafe(true).build();
+          CallOptions callOptions = CallOptions.DEFAULT;
+          if (noCacheCheckBox.isChecked()) {
+            callOptions =
+                callOptions.withOption(SafeMethodCachingInterceptor.NO_CACHE_CALL_OPTION, true);
+          }
+          if (onlyIfCachedCheckBox.isChecked()) {
+            callOptions =
+                callOptions.withOption(
+                    SafeMethodCachingInterceptor.ONLY_IF_CACHED_CALL_OPTION, true);
+          }
+          reply =
+              ClientCalls.blockingUnaryCall(
+                  channelToUse, safeCacheableUnaryCallMethod, callOptions, message);
+        } else {
+          GreeterGrpc.GreeterBlockingStub stub = GreeterGrpc.newBlockingStub(channelToUse);
+          reply = stub.sayHello(message);
+        }
+        return reply.getMessage();
+      } catch (Exception e) {
+        StringWriter sw = new StringWriter();
```

Android uses LOG right?

---

This is creating the string that gets displayed within the app, not writing to logcat. This is the approach used by both of our existing Android example apps: would you like to see something different here?

---

Well, I would expect to see LOG, or let the exception propagate up.  I don't think the existing android examples were written with best practices.  

That said, I would have just expected the exception to get thrown.

---

Added log. We'll get an exception here when e.g. the connection is UNAVAILABLE. This is printed to the screen, and now also logged.